riverdriver/riverpgxv5/internal/dbsqlc/river_job.sql

```diff
@@ -276,9 +278,43 @@ INSERT INTO river_job(
     @queue,
     coalesce(sqlc.narg('scheduled_at')::timestamptz, now()),
     @state,
-    coalesce(@tags::varchar(255)[], '{}')
+    coalesce(@tags::varchar(255)[], '{}'),
+    @unique_key
 ) RETURNING *;
 
+-- name: JobInsertUnique :one
+INSERT INTO river_job(
+    args,
+    created_at,
+    finalized_at,
+    kind,
+    max_attempts,
+    metadata,
+    priority,
+    queue,
+    scheduled_at,
+    state,
+    tags,
+    unique_key
+) VALUES (
+    @args,
+    coalesce(sqlc.narg('created_at')::timestamptz, now()),
+    @finalized_at,
+    @kind,
+    @max_attempts,
+    coalesce(@metadata::jsonb, '{}'),
+    @priority,
+    @queue,
+    coalesce(sqlc.narg('scheduled_at')::timestamptz, now()),
+    @state,
+    coalesce(@tags::varchar(255)[], '{}'),
+    @unique_key
+)
+ON CONFLICT (kind, unique_key) WHERE unique_key IS NOT NULL
+    -- Something needs to be updated for a row to be returned on a conflict.
+    DO UPDATE SET kind = EXCLUDED.kind
+RETURNING *, (xmax != 0) AS unique_skipped_as_duplicate;
```

Comment:
This `(max != 0)` thing is kind of a well known upsert trick for figuring out whether the operation resulted in an insert or an update. It's not formally documented, but it's worked since the introduction of upsert, and general consensus is that it's unlikely to be broken at this point. Unfortunately, upsert provides no formal way of differentiating an insert versus an update.

Reply:
Oh wild, nice trick!